libcontainer/cgroups/fs2/create.go

```diff
@@ -10,24 +11,27 @@ import (
 	"github.com/opencontainers/runc/libcontainer/configs"
 )
 
-// neededControllers returns the string to write to cgroup.subtree_control,
-// containing the list of controllers to enable (for example, "+cpu +pids"),
+func supportedControllers(cgroup *configs.Cgroup) ([]byte, error) {
+	const file = UnifiedMountpoint + "/cgroup.controllers"
+	return ioutil.ReadFile(file)
+}
+
+// neededControllers returns the number of controllers to enable,
 // based on (1) controllers available and (2) resources that are being set.
 //
 // The resulting string does not include "pseudo" controllers such as
 // "freezer" and "devices".
-func neededControllers(cgroup *configs.Cgroup) ([]string, error) {
-	var list []string
+func neededControllers(cgroup *configs.Cgroup) (int, error) {
```

Comment:
Maybe return `bool` here as no one is interested in the number of controllers to be enabled, but rather whether any controllers need to be enabled or not.

Reply:
You can also optimize things by returning true as soon as there is a controller that needs to be enabled.

Something like

```go
￼	// 
￼	have := func(controller string) bool {
￼		if _, ok := avail[controller]; ok {
￼	￼	￼	return true
￼		}
￼		return false
￼	}

...

if isPidsSet(cgroup) && have("pids") {
￼		return true
}
if isMemorySet(cgroup) && have("memory") {
    return true
}
....
```
		add("memory")
	}
		add("pids")

Reply:
And maybe rename to `needAnyControllers()` or smth similar